source/dub/generators/ninja.d

```diff
@@ -0,0 +1,597 @@
+				settings.buildType,         /* 2: Build type */
+				settings.compiler.name,     /* 3: Used compiler name */
+				settings.config,            /* 4: Used configuration */
+			);
+
+			ninjaBuilder.addRule(
+				"CLEAN", [
+					"command": "/usr/bin/ninja $FILE_ARG -t clean $TARGETS",
+					"description": "Cleaning all built files..."
+				]);
+			ninjaBuilder.addBuildTarget(
+				"CLEAN", "clean",
+				["PHONY"], ["pool": "console"]);
+
+			ninjaBuilder.addRule(
+				"HELP", [
+					"command": "/usr/bin/ninja -t targets",
+					"description": "All primary targets available:"
+				]);
+			ninjaBuilder.addBuildTarget(
+				"HELP", "help",
+				["PHONY"], ["pool": "console"]);
+
+			ninjaBuilder.addRule(
+				"CUSTOM_COMMAND", [
+					"command": "$COMMAND",
+					"description": "$DESC",
+					"restat": "1"
+				]);
+
+			// phony target to force build everytime
+			ninjaBuilder.addPhonyTarget("PHONY", []);
+		}
+
+		auto ruleDCompilerFlags = settings.compiler.name == "gdc"
+				? "-fdeps=$out.deps"
+				: "-makedeps=$out.deps";
+
+		ninjaBuilder.addRule(
+			"D_COMPILER", [
+				"command": format!`$dc -c %1$s $dflags $target_dflags $DFLAGS -of=$out $in`(ruleDCompilerFlags),
+				"description": "Compiling D object $out",
+				"deps": "gcc",
+				"depfile": "$out.deps"
+			]);
+
+		ninjaBuilder.addRule(
+			"D_LINK", [
+				"command": `$dc -of=$out @$out.rsp $lflags $target_lflags $LFLAGS`,
+				"description": "Linking target $out",
+				"rspfile": "$out.rsp",
+				"rspfile_content": "$in",
+			]);
+
+		ninjaBuilder.addRule(
+			"AR", [
+				"command": `rm -f $out && ar crs $out $in`,
+				"description": "Archiving target $out",
+			]);
+
+		// setup buildir and rootdir variables
+		string rootdirStr;
+		{
+			auto rootdir = m_project.rootPackage.path;
+			rootdir.endsWithSlash = false;
+			rootdirStr = rootdir.toNativeString();
+		}
+
+		// rule to regenerate build.ninja file
+		ninjaBuilder.addRule(
+			"REGENERATE_BUILD", [
+				"command": format!"%-(%s %|%)%s %(%s%| %)"(
+					runByNinja ? ["DUB_NINJA_GENERATOR=1", "DUB_NINJA_REGENERATE=1"] : null,
+					thisExePath(),
+					Runtime.args[1..$].dup ~ format!"--root=%s"(escapeShellCommand(rootdirStr))
+				),
+				"description": "Regenerating build files.",
+				"generator": "1"
+			]
+		);
+		ninjaBuilder.addBuildTarget(
+			"REGENERATE_BUILD", "reconfigure",
+			["PHONY"],
+			["pool": "console"]);
+
+		ninjaBuilder.addVariables([
+			"builddir": format!"%s/%s/%s"(builddir, settings.buildType, m_project.rootPackage.name),
+			"rootdir": rootdirStr,
+			"ninjapath": this.ninjaBuildPath
+		]);
+
+		// setup dc, dflags and lflags variables according to the build type
+		string compilerWrapper = settings.platform.compilerBinary
+			.replace("ldc", "ldmd");
+
+		// to enable colors on ninja
+		string[] dflags = [
+			settings.platform.compiler == "gdc"
+				? "-fdiagnostics-color=always"
+				: "-color=on"
+		];
+		string[] lflags;
+
+		switch(settings.buildType)
```

Comment:
this switch should probably use the Compiler class instead of reinventing these values

Reply:
Thanks. There is a lot of stuff I'm not aware of about dub internals and this is one of the things. I'm going to fix this when I have some time to touch this.